demo for removing yarn workspaces

amitzur opened the pull request:
Removing yarn workspaces has several advantages:
1. Less magic in the workflow
2. Ability to validate package correctness WRT dependencies from the monorepo (code is run from `node_modules` folder of each package - resembling real-life scenario)
3. More easily install/uninstall dependencies
4. Less issues with aliases (like `selenium-webdriver-3` propagating to protractor)

It also has drawbacks:
1. Working with source code across packages requires manual `link`ing
2. Need to adjust our tooling and workflow
3. Updates go through the internet

I have motivation to closely check if we want to remove workspaces. So this PR is a demo for how it would look like and what issues we face.
The demo is only applied to `eyes-webdriverio-5` package.

Points to notice:
1. I had to publish to npm the packages `sdk-release-kit` and `sdk-fake-eyes-server@2.0.0`.
2. Workflow includes `yarn link` and `yarn unlink`. We can develop scripts to do that, as well as detect what is linked. We can do it granular or binary - link the entire monorepo or unlink it. It's up to us to decide on the workflow.
3. Upgrading deps within the monorepo is done with `yarn upgrade --exact --latest --pattern "@applitools/"`. This replaces `npx bongo verify-versions --fix`.
4. There's no more root-level package.json
5. `/scripts` folder moved to `sdk-shared/src/scripts`
6. `bongo preversion` hook now also yarn installs and upgrades

The major point to discuss is that now there is a new operation to execute. After pulling from github, developers need to run `yarn` or add/remove some linking.
Ultimately, this is possible because we can work on each package independently. Core today has enough tests to be able to be published on its own. If coverage tests then fail we can republish a core version, but this will probably not happen, as we will test it before when linked.

Thoughts?

tourdedave commented:
My initial thoughts:

- I'm a +1 on this
- We'll want a template for adding new packages to the mono (obv.)
- re: relying on an internet connection -- yarn supports an offline mirror ([link](https://classic.yarnpkg.com/blog/2016/11/24/offline-mirror/)). I'm not advocating for it, but food for thought.
- There will be friction supporting packages that live in repos outside of the mono, but this change will help.

amitzur commented:
@kyryloonufriiev waiting to hear your thoughts

kyryloonufriiev commented:
I'm on board, let's do it.

amitzur commented:
Another implication of this is that currently private utility packages (like sdk-shared) need to be published when changes are done. This means also coverage tests. We can do something to workaround that if we want. Just a thought - a file like `supported-tests` which doesn't have any dependencies could be read from file system instead of the pacakge.

kyryloonufriiev commented:
Couldn't we just `link` whole `sdk-shared` package?

tourdedave commented:
Agreed. For devDependencies that are private, I don’t see a good reason to
make them public when we can link them (or do something else). Unless we
want to offer the choice as part the developer experience (linking or rely
on a public registry).

On Sun, Aug 9, 2020 at 01:32 Kyrylo <notifications@github.com> wrote:

> Couldn't we just link whole sdk-shared package?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/applitools/eyes.sdk.javascript1/pull/232#issuecomment-671008980>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADPTMMCTWOECGK35VISSJ3R7YYFDANCNFSM4PSMB3IA>
> .
>

tourdedave commented:
I forgot about the use case of packages not in our mono that we want to use
internal tooling on. In that case, making these packages public is a
good idea.

On Sun, Aug 9, 2020 at 08:13 Dave Haeffner <dave.haeffner@applitools.com>
wrote:

> Agreed. For devDependencies that are private, I don’t see a good reason to
> make them public when we can link them (or do something else). Unless we
> want to offer the choice as part the developer experience (linking or rely
> on a public registry).
>
> On Sun, Aug 9, 2020 at 01:32 Kyrylo <notifications@github.com> wrote:
>
>> Couldn't we just link whole sdk-shared package?
>>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/applitools/eyes.sdk.javascript1/pull/232#issuecomment-671008980>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AADPTMMCTWOECGK35VISSJ3R7YYFDANCNFSM4PSMB3IA>
>> .
>>
>

amitzur commented:
ready for review!

Notable changes:
1. Each package now has its own eslint and eslintrc. The ones that didn't already have one got their copy from the master .eslintrc.
2. Prettier is version 1.19, but I saw that version 2 has nicer formatting so going forward I'd like to upgrade to that.
3. Each package's `dependencies` are fixed, including non-applitools packages.
4. `devDependencies` are not fixed.
5. The command to upgrade dependencies is `yarn deps`. The reason it's not just `yarn upgrade` is because `dependencies` are fixed. This command also validates that dependencies are fixed, and also commits package.json, yarn.lock, and CHANGELOG.md. By default it only upgrades applitools packages from `dependencies`, and all packages from `devDependencies`. 
  5a. `yarn deps --upgrade-all` will upgrade also non-applitools packages.
  5b. `yarn deps --skip-dev` will skip upgrading devDependencies.
  5c. `yarn deps --skip-commit` will not commit the changed files.
6. I changed the auto commit message to `[auto commit] ${pkgName}: upgrade deps`
7. In the preversion check, `verifyVersions` is now a warning, not an error
8. In the preversion check, `verifyCommits` is now a warning, not an error
9. There were existing flags for preversion as env vars, and I changed them to flags: `--skip-verify-commits` and `--skip-verify-installed-versions`. I needed them for a very unusual case once and I don't think we will use them.

---

What I will do after the merge:
1. publish @applitools/sdk-release-kit
2. run `yarn deps` on all packages

What everyone needs to do after the merge:
```
rm -rf node_modules
rm -rf packages/*/node_modules
```

kyryloonufriiev commented:
Looks really good to me! Great job.
The only question is why do we want to have fixed versions? Just to be safe of possibly breaking changes in minor versions?

tourdedave reviewed (approved):
LGTM - ship it shabtay

tourdedave commented:
> The only question is why do we want to have fixed versions? Just to be safe of possibly breaking changes in minor versions?

The initial rationale was two-fold:

1) To protect against `instanceOf` exceptions (which can be grouped under `possibly breaking changes in minor versions`). Fixed versions an our pack-install pre-release check rendered this a non-issue. 

2. To know what the customer is using without needing any extra round trips for that information. That, and the added benefit of being able to reproduce the exact SDK version without needing to jump through any hoops.

amitzur commented:
Okay, I started to write an answer and Dave beat me to it. That's exactly the answer.

My addition:
True, I'd hate it if we needed to release an SDK version in order to ship a bug fix in a 3rd party. But that's rare, and I favor the consistency.

amitzur commented on packages/eyes-cypress/.eslintignore:
```diff
@@ -1,2 +1,8 @@
 /commands.js
-**/fixtures
+test/fixtures/testApp/cypress/support
```
Nope. Good points, I'll change it

amitzur commented on packages/eyes-protractor/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 # Change Log
 
 ## Unreleased
+- updated to @applitools/eyes-sdk-core@12.2.5 (from 12.2.4)
+- updated to @applitools/visual-grid-client@15.0.5 (from 15.0.4)
```
10x
